File: OpenWrap/Basic/DFP/ObjectiveC/SampleApp/SampleApp/EventHandlers/DFP/Interstitial/DFPInterstitialEventHandler.m
//This class is compatible with OpenWrap SDK v1.5.0

#import "DFPInterstitialEventHandler.h"

#define SYNC_TIMEOUT_INTEREVAL 0.8

@interface DFPInterstitialEventHandler () <
GADAppEventDelegate,
GADInterstitialDelegate>
@property(nonatomic, strong) DFPInterstitial *interstitial;
@property (nonatomic, weak) id<POBInterstitialEventDelegate> delegate;
@property (nonatomic, copy) NSString *adUnitId;
@property(nonatomic) NSTimer *timer;
@property (nonatomic, assign) BOOL notified;
@property(nonatomic, assign) BOOL isAppEventExpected;
@end

@implementation DFPInterstitialEventHandler
- (instancetype)initWithAdUnitId:(NSString *)adUnitId {
    
    self = [super init];
    if (self) {
        _adUnitId = adUnitId;
    }
    return self;
}

- (void)dealloc {
    _interstitial.delegate = nil;
    _interstitial = nil;
    _configBlock = nil;
}

#pragma mark - POBInterstitialEvent methods

- (void)setDelegate:(id<POBInterstitialEventDelegate>)delegate {
    _delegate = delegate;
}

- (void)requestAdWithBid:(POBBid *)bid {
    
    _notified = NO;
    _isAppEventExpected = NO;
    
    // Create DFPInterstitial and set ad unit
    _interstitial = [[DFPInterstitial alloc]
                     initWithAdUnitID:self.adUnitId];
    
    // Set delegates on DFPInterstitial instance, these should not be removed/overridden else event handler will not work as expected.
    _interstitial.delegate = self;
    _interstitial.appEventDelegate = self;

    // Create DFP ad request
    DFPRequest *dfpRequest = [[DFPRequest alloc] init];
    
    // Call configuration block if set. it can be used to configure DFP banner and ad request
    if (self.configBlock) {
        self.configBlock(_interstitial, dfpRequest);
    }
    
    if (!(_interstitial.appEventDelegate == self ||
          _interstitial.delegate == self)) {
        NSLog(@"Do not set DFP delegates. These are used by DFPInterstitialEventHandler internally.");
    }

    // If bid is valid, add bid related custom targetting on DFP ad request
    if (bid) {
        if (bid.status.boolValue) {
            _isAppEventExpected = YES;
        }
        NSMutableDictionary * customTargeting = [NSMutableDictionary dictionaryWithDictionary:dfpRequest.customTargeting];
        [customTargeting addEntriesFromDictionary:[bid targetingInfo]];
        [dfpRequest setCustomTargeting:[NSDictionary dictionaryWithDictionary:customTargeting]];
        NSLog(@"Custom targeting : %@", [customTargeting description]);
    }
    
    // Load ad request
    [_interstitial loadRequest:dfpRequest];
}

-(void)showFromViewController:(UIViewController *)controller{
    [self.interstitial presentFromRootViewController:controller];
}

#pragma mark - GADAppEventDelegate methods
/// Called when the interstitial receives an app event.
- (void)interstitial:(GADInterstitial *)interstitial
  didReceiveAppEvent:(NSString *)name
            withInfo:(nullable NSString *)info{
    
    if ([name isEqualToString:@"pubmaticdm"]) {
        if (_notified) {
            NSDictionary *userInfo = nil;
            NSString *errorMessage = @"App event is called unexpetedly";
            userInfo = @{NSLocalizedDescriptionKey : NSLocalizedString(errorMessage, nil)};
            NSError *error = [NSError errorWithDomain:kPOBErrorDomain
                                                 code:POBSignalingError
                                             userInfo:userInfo];
            [self.delegate failedWithError:error];
            return;
        }
        _notified = YES;
        _interstitial.delegate = nil;
        _interstitial = nil;
        [self.delegate openWrapPartnerDidWin];
    }
}

#pragma mark - GADInterstitialDelegate methods
- (void)interstitialDidReceiveAd:(GADInterstitial *)ad{
    if (_notified) {
        return;
    }
    
    // If OpenWrap SDK have provided non-zero bid price, expect for app event for fixed time interval, otherwise consider as DFP has won & serving its own ad
    if (!_isAppEventExpected) {
        if (!self.notified) {
            [self.delegate adServerDidWin];
            self.notified = YES;
        }
    } else {
        // Timer to synchronize did recieve and app event callback as their sequence is not fixed
        [self.timer invalidate];
        self.timer = [NSTimer scheduledTimerWithTimeInterval:SYNC_TIMEOUT_INTEREVAL target:self selector:@selector(syncTimetAction) userInfo:nil repeats:NO];
    }
}

- (void)interstitial:(GADInterstitial *)ad didFailToReceiveAdWithError:(GADRequestError *)error{
    NSError *eventError = nil;
    switch (error.code) {
        case kGADErrorNoFill:
            eventError = [self eventError:error withErrorCode:POBErrorNoAds];
            break;
            
        case kGADErrorInvalidRequest:
            eventError = [self eventError:error withErrorCode:POBErrorInvalidRequest];
            break;
            
        case kGADErrorNetworkError:
            eventError = [self eventError:error withErrorCode:POBErrorNetworkError];
            break;
            
        case kGADErrorTimeout:
            eventError = [self eventError:error withErrorCode:POBErrorTimeout];
            break;
            
        case kGADErrorInternalError:
        case kGADErrorInterstitialAlreadyUsed:
        case kGADErrorMediationDataError:
        case kGADErrorMediationAdapterError:
        case kGADErrorMediationInvalidAdSize:
        case kGADErrorInvalidArgument:
            eventError = [self eventError:error withErrorCode:POBErrorInternalError];
            break;
            
        case kGADErrorReceivedInvalidResponse:
            eventError = [self eventError:error withErrorCode:POBErrorInvalidResponse];
            break;
            
        default:
            eventError = error;
            break;
    }
    
    [_delegate failedWithError:eventError];
}

-(NSError *)eventError:(GADRequestError *)error  withErrorCode:(POBErrorCode )code{
    NSError *eventError = [NSError errorWithDomain:kPOBErrorDomain
                                              code:code
                                          userInfo:error.userInfo];
    return eventError;
}

- (void)interstitialWillPresentScreen:(GADInterstitial *)ad{
    [self.delegate willPresentAd];
}

- (void)interstitialDidDismissScreen:(GADInterstitial *)ad{
    [self.delegate didDismissAd];
}

- (void)interstitialWillLeaveApplication:(GADInterstitial *)ad{
    /*
     GADInterstitialDelegate doesn't provide didClickAd callback. Since user click
     has triggered the willLeaveApplication event, the interstitial event handler
     can safely give didClickAd callback.
     */
    [self.delegate didClickAd];
    [self.delegate willLeaveApp];
}
#pragma mark -

- (void)syncTimetAction{
    if (!self.notified) {
        [self.delegate adServerDidWin];
        self.notified = YES;
    }
}

@end

File: OpenBid/DFP/Swift/SampleApp/SampleApp/EventHandlers/DFP/Interstitial/DFPInterstitialEventHandler.swift


import UIKit
import GoogleMobileAds

/*!
 This class is responsible for communication between OpenBid interstitial and DFP interstitial .
 It implements the POBInterstitialEvent protocol. it notifies event back to OpenBid SDK using POBInterstitialEventDelegate methods
 */
class DFPInterstitialEventHandler: NSObject,POBInterstitialEvent,GADAppEventDelegate,GADInterstitialDelegate {

    let SYNC_TIMEOUT_INTEREVAL = 0.8
    
    var interstitial: DFPInterstitial?
    weak var _delegate: POBInterstitialEventDelegate?
    var adUnitId = ""
    var timer: Timer?
    var notified = false
    var isAppEventExpected = false
    
    /*!
     A configBlock that is called before event handler makes ad request call to DFP SDK. It passes DFPInterstitial & DFPRequest which will be used to make ad request.
     */
    var configBlock: ((_ view: DFPInterstitial, _ request: DFPRequest)->Void)?

    init(_ adUnitId: String) {
        super.init()
        self.adUnitId = adUnitId
    }
    
    deinit {
        interstitial?.delegate = nil
        interstitial = nil
    }
    
    func setDelegate(_ delegate: POBInterstitialEventDelegate?) {
        _delegate = delegate
    }

    func requestAd(with bid: POBBid?) {
        
        notified = false
        isAppEventExpected = false
        interstitial = DFPInterstitial(adUnitID: adUnitId)
        interstitial?.delegate = self
        interstitial?.appEventDelegate = self
        
        let dfpRequest = DFPRequest()
        
        self.configBlock?(interstitial!,dfpRequest)
        
        if !(interstitial?.appEventDelegate is DFPInterstitialEventHandler
            && interstitial?.delegate is DFPInterstitialEventHandler) {
            NSLog("Do not set DFP delegates. These are used by DFPInterstitialEventHandler internally.");
        }

        if bid != nil {
            if bid?.status.boolValue ?? false {
                isAppEventExpected = true
            }
            let customTargeting = NSMutableDictionary()
            customTargeting.addEntries(from: dfpRequest.customTargeting ?? [:])
            customTargeting.addEntries(from: bid?.targetingInfo() ?? [:]);
            dfpRequest.customTargeting = customTargeting as? [AnyHashable : Any]
            print("Custom targeting : \(String(describing: customTargeting.description))")
        }
        interstitial?.load(dfpRequest)
    }
    
    func show(from controller: UIViewController) {
        interstitial?.present(fromRootViewController: controller)
    }

    func interstitial(_ interstitial: GADInterstitial, didReceiveAppEvent name: String, withInfo info: String?) {
        
        if (name == "pubmaticdm") {            
            if notified {
                var userInfo: [AnyHashable : Any]? = nil
                let errorMessage = "App event is called unexpetedly"
                userInfo = [NSLocalizedDescriptionKey: NSLocalizedString(errorMessage, comment: "")]
                let error = NSError(domain: kPOBErrorDomain, code: POBErrorCode.signalingError.rawValue, userInfo: userInfo as? [String : Any])
                _delegate?.failedWithError(error)
                return
            }
            notified = true
            self.interstitial?.delegate = nil
            self.interstitial = nil
            _delegate?.openBidPartnerDidWin()
        }
    }
    
    func interstitialDidReceiveAd(_ ad: GADInterstitial) {
        if notified {
            return
        }
        
        if !isAppEventExpected {
            if !notified {
                _delegate?.adServerDidWin()
                notified = true
            }
        } else {
            // Timer to synchronize did recieve and app event callback as their sequence is not fixed
            timer?.invalidate()
            timer = Timer.scheduledTimer(timeInterval: SYNC_TIMEOUT_INTEREVAL, target: self, selector: #selector(self.syncTimetAction), userInfo: nil, repeats: false)
        }
    }
    
    @objc func syncTimetAction() {
        if !notified {
            _delegate?.adServerDidWin()
            notified = true
        }
    }
    
    func interstitial(_ ad: GADInterstitial, didFailToReceiveAdWithError error: GADRequestError) {
        
        var eventErr:NSError?
        switch error.code {
        case GADErrorCode.noFill.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorNoAds)
            
        case GADErrorCode.invalidRequest.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorInvalidRequest)
            
        case GADErrorCode.networkError.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorNetworkError)
            
        case GADErrorCode.timeout.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorTimeout)
            
        case GADErrorCode.internalError.rawValue:
            fallthrough
        case GADErrorCode.interstitialAlreadyUsed.rawValue:
            fallthrough
        case GADErrorCode.mediationDataError.rawValue:
            fallthrough
        case GADErrorCode.mediationNoFill.rawValue:
            fallthrough
        case GADErrorCode.invalidArgument.rawValue:
            fallthrough
        case GADErrorCode.mediationAdapterError.rawValue:
            fallthrough
        case GADErrorCode.mediationInvalidAdSize.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorInternalError)

        case GADErrorCode.receivedInvalidResponse.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorInvalidResponse)
            
        default:
            eventErr = error
            break
        }
        _delegate?.failedWithError(eventErr)
    }
    
    func eventError(_ error : GADRequestError, code : POBErrorCode) -> NSError {
        let eventError = NSError(domain: kPOBErrorDomain, code:code.rawValue, userInfo: error.userInfo)
        return eventError
    }
    
    func interstitialWillPresentScreen(_ ad: GADInterstitial) {
        _delegate?.willPresentAd()
    }
    
    func interstitialDidDismissScreen(_ ad: GADInterstitial) {
        _delegate?.didDismissAd()
    }
    
    func interstitialWillLeaveApplication(_ ad: GADInterstitial) {
        /*
         GADInterstitialDelegate doesn't provide didClickAd callback. Since user click
         has triggered the willLeaveApplication event, the interstitial event handler
         can safely give didClickAd callback.
         */
        _delegate?.didClickAd()
        _delegate?.willLeaveApp()
    }
}
